stan/math/prim/mat/fun/gp_dot_prod_cov.hpp

```diff
@@ -15,142 +16,308 @@ namespace stan {
 namespace math {
 
 /**
- * Returns a dot product covariance matrix. A member of Stan's Gaussian Process
- * Library.
- *
- * \f$k(x,x') = \sigma^2 + x \cdot x'\f$
+ * Returns a dot product covariance matrix. A member of Stan's Gaussian
+ * Process Library.
  *
- * A dot product covariance matrix is the same covariance matrix
- * as in bayesian regression with \f$N(0,1)\f$ priors on regression coefficients
- * and a \f$N(0,\sigma^2)\f$ prior on the constant function. See Rasmussen and
- * Williams et al 2006, Chapter 4.
+ * \f$k(x,x') = \sigma^2 x \cdot x'\f$
  *
  * @tparam T_x type of std::vector of elements
- * @tparam T_sigma type of sigma
+ * @tparam T_sigma_squared scalar type of sigma_squared
  *
- * @param x std::vector of elements that can be used in dot product.
+ * @param x std::vector of elements that can be used in transpose
+ *   and multiply
  *    This function assumes each element of x is the same size.
- * @param sigma constant function that can be used in stan::math::square
+ * @param sigma_squared variance
  * @return dot product covariance matrix that is positive semi-definite
  * @throw std::domain_error if sigma < 0, nan, inf or
  *   x is nan or infinite
  */
-template <typename T_x, typename T_sigma>
-Eigen::Matrix<return_type_t<T_x, T_sigma>, Eigen::Dynamic, Eigen::Dynamic>
+template <typename T_x, typename T_sigma_squared>
+Eigen::Matrix<return_type_t<T_x, T_sigma_squared>, Eigen::Dynamic,
+              Eigen::Dynamic>
+gp_dot_prod_cov(const std::vector<T_x> &x,
+                const T_sigma_squared &sigma_squared) {
+  check_not_nan("gp_dot_prod_cov", "sigma_squared", sigma_squared);
+  check_positive_finite("gp_dot_prod_cov", "sigma_squared", sigma_squared);
+
+  size_t x_size = x.size();
+  check_not_nan("gp_dot_prod_cov", "x", x);
+  check_finite("gp_dot_prod_cov", "x", x);
+
+  Eigen::Matrix<return_type_t<T_x, T_sigma_squared>, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+  if (x_size == 0) {
+    return cov;
+  }
+
+  for (size_t i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_squared * x[i] * x[i];
+    for (size_t j = i + 1; j < x_size; ++j) {
+      cov(i, j) = sigma_squared * x[i] * x[j];
+      cov(j, i) = cov(i, j);
+    }
+  }
+  cov(x_size - 1, x_size - 1) = sigma_squared * x[x_size - 1] * x[x_size - 1];
+  return cov;
+}
+
+/**
+ * Returns a dot product covariance matrix. A member of Stan's Gaussian
+ * Process Library.
+ *
+ * \f$k(x,x') = \sigma^2 x \cdot x'\f$
+ *
+ * @tparam T_x1 type of first std::vector of double
+ * @tparam T_x2 type of second std::vector of double
+ * @tparam T_sigma_squared scalar type of sigma_squared
+ *
+ * @param x1 std::vector of elements that can be used in dot_product
+ * @param x2 std::vector of elements that can be used in dot_product
+ * @param sigma_squared variance
+ * @return dot product covariance matrix
+ * @throw std::domain_error if sigma < 0, nan or inf
+ *   or if x1 or x2 are nan or inf
```

Comment:
No big deal, but we could have two separate `@throw` statements, one for sigma and one for x1, x2.